src/blib2to3/pgen2/parse.py

```diff
@@ -54,21 +65,45 @@ def __init__(self, parser: "Parser", ilabels: List[int], context: Context) -> No
 
         self._dead_ilabels: Set[int] = set()
         self._start_point = self.parser.stack
-        self._points = {ilabel: copy.deepcopy(self._start_point) for ilabel in ilabels}
+        self._points = {ilabel: stack_copy(self._start_point) for ilabel in ilabels}
 
     @property
     def ilabels(self) -> Set[int]:
         return self._dead_ilabels.symmetric_difference(self._ilabels)
 
     @contextmanager
     def switch_to(self, ilabel: int) -> Iterator[None]:
-        self.parser.stack = self._points[ilabel]
+        with self.patch():
+            self.parser.stack = self._points[ilabel]
+            try:
+                yield
+            except ParseError:
+                self._dead_ilabels.add(ilabel)
+            finally:
+                self.parser.stack = self._start_point
+
+    @contextmanager
+    def patch(self) -> Iterator[None]:
+        """
+        Patch basic state operations (push/pop/shift) with node-level
+        immutable variants. These still will operate on the stack; but
+        they won't create any new nodes, or modify the contents of any
+        other existing nodes.
+
+        This saves us a ton of time when we are backtracking, since we
+        want to restore to the initial state as quick as possible, which
+        can only be done by having as little mutatations as possible.
+        """
+        original_functions = {}
+        for name in self.parser.STATE_OPERATIONS:
+            original_functions[name] = getattr(self.parser, name)
+            safe_variant = getattr(self.parser, name + "_safe")
+            setattr(self.parser, name, safe_variant)
```

Comment:
Is mypyc going to like this?

Reply:
I was unsure about this one, but at worst we can simply merge the implementations of the safe/unsafe and guard the safe one with a flag like `parser.is_safe = True`. CC: @ichard26